### PR TITLE
Unify donor validation code

### PR DESCRIPTION
**Pull request by JeroenDeDauw**
https://phabricator.wikimedia.org/T180936

Plan & status:

* [x] Split donor validation out of AddDonationValidator
* [x] Kill DonorValidator, DonorAddressValidator and DonorNameValidator in Donation BC
* [x] Add dedicated tests for new UC
* [ ] [Remove usage of killed code from FF, remove logic in route handler, use new UC](https://github.com/wmde/FundraisingFrontend/pull/1041)

### CODE REVIEW

**Comment by wiese** [comment]
Ah. So it was
https://github.com/wmde/fundraising-donations/pull/27/files#diff-f0103fc548b80e63f07de4f8a6cc309dL234 vs https://github.com/wmde/fundraising-donations/pull/27/files#diff-0f40232f1ceac1281dfd5c50cfa28f4bL26

**Comment by JeroenDeDauw** [comment]
That is one difference yes: the validate-address route did not check the format of the post code. Another difference is that the AddDonationUC checks for the max length of the fields while the forementioned route did not.

File: src/UseCases/AddDonation/AddDonationValidator.php [hunk 1]
```diff
@@ -51,7 +44,7 @@ public function __construct( PaymentDataValidator $paymentDataValidator, BankDat
 
 		$this->paymentDataValidator = $paymentDataValidator;
 		$this->bankDataValidator = $bankDataValidator;
-		$this->emailValidator = $emailValidator;
+		$this->donorValidator = new ValidateDonorUseCase( $emailValidator );
```
Comment (gbirke) [inline src/UseCases/AddDonation/AddDonationValidator.php]: I'm confused by this line. Use cases initializing use cases? Is that allowed/possible? What *is* a use case, conceptually? Let's talk about this.

File: src/UseCases/ValidateDonor/ValidateDonorRequest.php [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	private $city;
+	private $countryCode;
+	private $emailAddress;
+
+	public static function newInstance(): self {
+		return new self();
+	}
+
+	public function withType( string $donorType ): self {
+		$request = clone $this;
```
Comment (gbirke) [inline src/UseCases/ValidateDonor/ValidateDonorRequest.php]: I like this fluent interface. It's a bit confusing when you first see the class, but when you see the calls, it reads very well, which makes up for the confusion. 

What is the reasoning behind the cloning? It creates a kind of "immutable class". I'm interested if something in the interface made this necessary or if you just considered it good style.

Comment (JeroenDeDauw) [reply src/UseCases/ValidateDonor/ValidateDonorRequest.php]: Good style. This is yet another hack to create a ValueObject or DTO in PHP...

File: src/UseCases/AddDonation/AddDonationValidator.php (hunk 1 above)
Comment (JeroenDeDauw) [reply src/UseCases/AddDonation/AddDonationValidator.php]: Yes lets talk

**Pull request merged**